crates/forge/src/init.rs

```diff
@@ -0,0 +1,122 @@
+use anyhow::{bail, Result};
+
+use include_dir::{include_dir, Dir, DirEntry};
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+static TEMPLATE: Dir = include_dir!("starknet_forge_template");
+
+fn create_with_template(dir: &Dir<'_>, base_path: &PathBuf, project_name: &str) -> Result<()> {
+    fs::create_dir_all(base_path)?;
+
+    for entry in dir.entries() {
+        let path = base_path.join(entry.path());
+
+        match entry {
+            DirEntry::Dir(d) => {
+                fs::create_dir_all(&path)?;
+                create_with_template(d, base_path, project_name)?;
+            }
+            DirEntry::File(f) => {
+                let contents = f.contents();
+                let contents = replace_project_name(contents, project_name);
+
+                fs::write(path, contents)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn replace_project_name(contents: &[u8], project_name: &str) -> Vec<u8> {
+    // SAFETY: We control these files, and we know that they are UTF-8.
+    let contents = unsafe { std::str::from_utf8_unchecked(contents) };
+    let contents = contents.replace("{{ PROJECT_NAME }}", project_name);
+    contents.into_bytes()
+}
+
+fn check_path(path: &Path) -> Result<()> {
+    // warn if the path contains characters that will break `env::join_paths`
+    if std::env::join_paths(std::slice::from_ref(&OsStr::new(path))).is_err() {
+        let path = path.to_string_lossy();
+        bail!(format!(
+            "the path `{path}` contains invalid PATH characters (usually `:`, `;`, or `\"`)\n\
+            It is recommended to use a different name to avoid problems."
+        ));
+    }
+    Ok(())
+}
```

Comment:
Wouldn't using https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.canonicalize  or  this and checking for errors be enough of path validation? 

I'm not sure if this kind of validation is a good/necessary approach.

Reply:
Strong +1

Reply:
I realize that I validate it double, because I also check if `characters are ASCII lowercase letters, ASCII numbers or underscore"`